File: CS50X/1. C/credit/credit.c
// Given a credit card number, checks if it is valid and returns which company issued it

#include <cs50.h>
#include <stdio.h>
#include <math.h>

int number_of_digits (long number);
int luhn (long card_number);
int first_two_digits (long number);
int get_xth_digit_from_last (long number, int x);

int main(void)
{
    long card_number = get_long ("Number: ");
    int n_digits = number_of_digits (card_number);
    int first_two = first_two_digits (card_number);
    int checksum = luhn (card_number);

    //AMEX
    if (checksum == 0 && n_digits == 15 && (first_two == 34 || first_two == 37))
    {
        printf ("AMEX\n");
    }

    //MASTERCARD
    else if (checksum == 0 && n_digits == 16 && first_two <=55 && first_two >=51)
    {
        printf ("MASTERCARD\n");
    }

    //VISA
    else if (checksum == 0 && (n_digits == 13 || n_digits == 16) && first_two >=40 && first_two <=49)
    {
        printf ("VISA\n");
    }

    else
    {
        printf ("INVALID\n");
    }
}

int luhn (long card_number) //calculates the checksum for a card number
{
    int n_digits = number_of_digits (card_number);
    int even = 0;
    int odd = 0;
    for (int i=2; i<=n_digits; i+=2) //goes to every even numbered digit (counting from the right)
    {
        int x = 2 * get_xth_digit_from_last (card_number, i);
        switch (x) // Since there are only a few cases, I do this manually.
            {
            case 10:
                x = 1;
                break;
            case 12:
                x = 3;
                break;
            case 14:
                x = 5;
                break;
            case 16:
                x = 7;
                break;
            case 18:
                x = 9;
                break;
            default:
                break;
        }
        even = even + x; //adds the latest digit in the for loop to a rolling sum.
    }

    for (int i=1; i<=n_digits; i+=2) //does the logic for the odd-numbered digits (counting from the right)
    {
        int x = get_xth_digit_from_last (card_number, i);
        odd = odd + x;
    }

    return (odd+even) % 10; //this takes the last digit of the sum, which is the only thing that's relevant.
}

//Counts number of digits in a long integer.
int number_of_digits (long number)
{
    int n_digits=0;
    while (number>=1)
    {
        n_digits++;
        number/=10; //effectively removes the right-most digit  - since fractions aren't supported (just truncated)
    }
    return n_digits;
}

//returns the first 2 digits of a long integer
int first_two_digits (long number)
{
    while (number >=100)
    {
        number = number/10; //effectively removes the right-most digit
    }
    return number;
}

//Gets the xth digit from the right
int get_xth_digit_from_last (long number, int x)
{
    for (int i=0; i<x-1; i++)
    {
        number = number/10; //removes the last digit.
    }
    number = number % 10; //removes all the preceding digits.
    return number;
}

File: CS50X/6. Python/6sentimental-credit/credit.py
from cs50 import get_string
import re


def main():
    card_number = get_string("Number: ")
    length = len(card_number)
    checksum = luhn(card_number)

    # Set up regular expressions for each of the 3 card number formats, and check if they match the card number given.
    # Also check the checksum

    if checksum != 0:
        print("INVALID")
    elif re.match('3[4,7]\d{13}$', card_number):
        print("AMEX")
    elif re.match('5[1-5]\d{14}$', card_number):
        print("MASTERCARD")
    elif re.match('4(\d{12}|\d{15})$', card_number):
        print("VISA")
    else:
        print("INVALID")


def luhn(number):
    """Calculates checksum based on Luhn's algorithm for a credit card number"""
    even_sum = 0
    length = len(number)

    # Counts backward every 2 digits starting from the 2nd last digit of the card number
    for i in range(length-2, -1, -2):
        even_digit = 2 * int(number[i])
        # Replace even_digit with the sum of its digits if it is > 1 digit long
        if even_digit >= 10:
            even_digit = 1 + even_digit % 10
        even_sum += even_digit

    odd_sum = 0
    for i in range(length-1, -1, -2):
        odd_digit = int(number[i])
        odd_sum += odd_digit

    # Return the last digit of the number
    return (even_sum + odd_sum) % 10


main()